### ISSUE TITLE
Feedback Frontend

### ISSUE THREAD

**Issue by moedn** [wmde/LOSH-Frontend#58]
Hey there! I took a look onto the front page at https://zen-stonebraker-a93ddf.netlify.app/ and wanted to share my feedback :)

First: great work! Looks neat, well designed (keeps the O!N design) and even scales well on mobile devices.

2 small things:

- preferred columns for the table:
  - name
  - version
  - license
  - data-source
  - organisation
  - certificate
- button for CSV export of search results seems to be missing

Since we're importing data from different platforms (which appears to be quite messy), we should only use fields for which data is mostly available (which is also why we went for the list view as e.g. OSHWA projects (~1600) never come with an image)

Should I already define some of those action buttons or is it too early for that?

---

**EDIT**

- added 'certificate' to columns

**Comment by botoxparty** [comment]
Hey @moedn .

Thanks for the feedback about the table columns, this is something we had been waiting on the data to finalise.

If you have some preferences for the action buttons in the filter please feel free to share them, I was working under the assumption that these would be informed by the available columns.

**Comment by moedn** [comment]
Alrighty, here are my suggestions for action buttons. Let me know which of those seem doable for you within your time frame. Also, I'm very open to suggestions if you think e.g. a crucial function is missing or some of my suggestions simply make no sense from your perspective :) ← just let me know

So, yes, all 'actions' should be doable by the information available from the columns. However, as for 'certified project` it would be cool to filter for projects that have _any_ certificate. filtering for licensing schemes would be super useful, but not essential for v1.0 of the LOSH

Filtering:

- [ ] licensing schemes (strong/weak/non copyleft) + specific licenses\
  (all licenses use SPDX identifiers, so we could easily sort them by those categories; however, SPDX does not classify licenses AFAIK so we might need to include a data field for that)
- [ ] certified project [bool] + specific certificates [OSHWA or DIN SPEC 3105]
- [ ] data source
  - = list of supported platforms (GitHub, GitLab, Appropedia, Wikifactory, OSHWA etc.)
- [ ] organisation

+ sort by any of those columns listed in my previous comment

**Comment by moedn** [comment]
also, (if it fits into the design) it would be super cool to have a explanatory subtitle on the front page, so people that don't know OPEN!NEXT or LOSH have a rough idea, what the thing does

something like:\
`Find Open Source Hardware matching your use case. LOSH crawls data from several platforms for you and connects it in a graph database. Go ahead and search for free/open designs by the search bar below`

**Comment by botoxparty** [comment]
@moedn Regarding the filtering and sorting functions. From my exploration of the API its only possible to sort by date created and date edited, but not by the values of properties.

With the filtering:

**Licensing schemes**
Can you please provide a list of the values that are used, and their associated classification

**Certified project**
A list of these should also be provided

**Data source**
There would need to be a property that indicates what the data source is. There is a `source` property but it seems to be unpopulated. The `repo` property contains a link to the repository but it can't be used to filter as the URL is different for each project.

**Organisation**
Is there a list of all possible organisations or is this a dynamic field?

**Comment by moedn** [comment]
Alrighty,

1. @toban @lucaswerkmeister it is not possible to sort things e.g. alphabetical from the values of a certain property?
2. license classification: sure; there are a few lists available online so I could worst case hardcode one by myself (still looking for a maintained list relying on SPDX)
  - Which would be your preferred format? CSV?
  - Do you want it in a git repo or via mail?
3. that's the data field `attestation`; I doubt that any of the current results have this field populated since the crawler doesn't get data from e.g. OSHWA yet. All items with a non-empty `attestation` field can be considered certified
4. there's a `datasource` field in the ontology ← is it unpopulated?
5. dynamic and even optional to fill out

**Comment by botoxparty** [comment]
2. License classification: Any format is fine, I would be expecting a list of valid values and their associated classification. I will then create a JSON object from this to use for mapping.
3. I don't see the `attestation` field on any of the current results.
4. I also don't `datasource` field populated on any results

**Comment by lucaswerkmeister** [comment]
1. If you’re getting the results from `wbsearchentities`, then there aren’t many sorting options available, you’d have to fetch all the results and sort them on your end. If you’re using SPARQL, you should be able to sort by anything you want.

**Comment by moedn** [comment]
to 

2. I started a CSV covering most popular licenses for OSH here: https://raw.githubusercontent.com/OPEN-NEXT/LOSH-Licenses/main/SPDX-reciprocality.csv
3. alright, so no real world data on that yet. Will come with the OSHWA crawl by the end of november
4. jo, that's indeed a bug; just had a meeting with the developer → created an [issue](https://github.com/ahane/LOSH-krawler/issues/2); he'll fix it
  - `dataSource` was ambiguous; we now added `repoHost`
    - `dataSource` = platform, where the crawled data is located
      - to be displayed only on the project detail page
    - `repoHost` = platform hosting the OSH documentation
      - takes the former position of `dataSource` as described above: filterable column

**Comment by moedn** [comment]
@botoxparty 

for

3. → do you need dummy data?
4. is this a blocker for the current version of the frontend? From my perspective it's not; this functionality can be added until the end of January

**Comment by botoxparty** [comment]
**2. Licence Classification**
So I was able to query for items classified as Weak and Strong but not Non-copyleft because of a character limitation when querying with cirrussearch.

**3. Certified project**
If we continue the filtering in this iteration, some dummy data would be good to test this is working as expected

**4. Data source**
The filter can't be added until the `repoHost` field exists and is populated. 

The queries may need to be written in SPARQL to have the desired filtering and sorting functionality.

@lucaswerkmeister Do you know if there is a character limit for SPARQL queries? Is it possible to do a full text search using SPARQL?

**Comment by lucaswerkmeister** [comment]
I don’t think there’s a length limit on SPARQL queries you’re realistically going to hit, but there is a length limit on GET request URLs, so at some point you’ll have to switch to POST for the queries. (See [T216590](https://phabricator.wikimedia.org/T216590).)

There’s not really a way to directly do a full-text search in SPARQL, but you might be able to use [MWAPI](https://www.mediawiki.org/wiki/Wikidata_Query_Service/User_Manual/MWAPI) to make a search request to the API.

**Comment by botoxparty** [comment]
@moedn @ckawohl-wmde 
It seems that the requirements for the filtering and sorting functionality cannot be fulfilled with the current implementation of search. 

SPARQL would need to be installed on the instance to investigate the possibilities of querying using SPARQL. Another potential option would be to query the elasticsearch instance directly but this would involve writing our own API. All these options fall outside the scope of the current sprint.

For now, I would recommend reviewing the detail page https://zen-stonebraker-a93ddf.netlify.app/detail/Q438

**Comment by moedn** [comment]
- To my knowledge, SPARQL is installed; queries can be submitted under https://losh.ose-germany.de/qs/
- :ok_hand: I'll review the detail page, thanks!

**Comment by lucaswerkmeister** [comment]
MWAPI only allows queries to certain hosts. It looks like [the Wikibase-Docker config](https://github.com/wmde/wikibase-release-pipeline/blob/7285d513afa92d2d51ac45030758dc67387dbc35/Docker/build/WDQS/mwservices.json#L105) is supposed to allowlist the Wikibase host by default, but maybe that’s not quite working… can the server admins find the `mwservices.json` inside the query service (backend) container and check which endpoints it allows?

**Comment by moedn** [comment]
to 2. → the [CSV](https://github.com/OPEN-NEXT/LOSH-Licenses/blob/main/SPDX-reciprocality.csv) should include now all licenses that will ever appear in crawls. There are a few more left on the SPDX list, but those may be a) pretty exotic and b) specific for software or other domains (such as fonts)

**Comment by aisbergg** [comment]
> MWAPI only allows queries to certain hosts. It looks like [the Wikibase-Docker config](https://github.com/wmde/wikibase-release-pipeline/blob/7285d513afa92d2d51ac45030758dc67387dbc35/Docker/build/WDQS/mwservices.json#L105) is supposed to allowlist the Wikibase host by default, but maybe that’s not quite working… can the server admins find the `mwservices.json` inside the query service (backend) container and check which endpoints it allows?

I applied a quick fix to the wdqs-backend service. The updater has trouble filling the db though. I guess Wikibase registered some special location `entity/statement`, which cannot be properly resolved with the current webserver configuration. I have to check this 

```plaintext
Unrecognized subjects:  [http://losh.ose-germany.de/entity/statement/Q436-9564CD2D-3AFE-4655-93D2-5708E8B2FA29, http://losh.ose-germany.de/entity/statement/Q436-6550A20B-DEC2-4F20-B782-176A869BDF9A, http://losh.ose-germany.de/entity/Q436, http://losh.ose-germany.de/entity/statement/Q436-485CACEC-255C-4867-BEE7-2CA8F94B3BDE, http://losh.ose-germany.de/index.php?title=Special:EntityData/Q436, http://losh.ose-germany.de/entity/statement/Q436-D0D217C4-77FF-4B80-8270-B97BEA22D8F1].  Expected only sitelinks and subjects starting with http://mediawiki.library-of-open-source-hardware.svc:8080/wiki/Special:EntityData/ and [http://mediawiki.library-of-open-source-hardware.svc:8080/entity/]
	at org.wikidata.query.rdf.tool.rdf.Munger$MungeOperation.finishCommon(Munger.java:800)
	at org.wikidata.query.rdf.tool.rdf.Munger$MungeOperation.munge(Munger.java:413)
	at org.wikidata.query.rdf.tool.rdf.Munger.munge(Munger.java:144)
	at org.wikidata.query.rdf.tool.Updater.handleChange(Updater.java:436)
	at org.wikidata.query.rdf.tool.Updater.lambda$fetchDataFromWikibaseAndMunge$7(Updater.java:286)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by lucaswerkmeister** [comment]
That sounds like Blazegraph’s `$WIKIBASE_HOST` isn’t set correctly – it’s expecting `http://mediawiki.library-of-open-source-hardware.svc:8080` instead of `http://losh.ose-germany.de` for some reason. (That could also explain why the correct host wasn’t allowed in MWAPI.)

**Comment by ckawohl-wmde** [comment]
This sounds to me like a clone/duplicate of the filter/search issue (#3)

**Comment by ckawohl-wmde** [comment]
@jakobw I think you might find some aspects of the filter/search issues were being discussed here as well.

**Comment by ckawohl-wmde** [comment]
This is a duplicate - closed because implementation is happening in #91
